File: python/151-200/161. One Edit Distance.py
class Solution(object):
    def isOneEditDistance(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        l1, l2 = len(s), len(t)
        if s == t or abs(l1 -l2) > 1:
            return False
        if l1 < l2:
            return self.isOneEditDistance(t, s)
        for i in range(l2):
            if s[i] != t[i]:
                return s[i+1:] == t[i:] or s[i+1:] == t[i+1:]
        return True

File: cpp/151-200/161. One Edit Distance.cc
#include <iostream>
#include <vector>
#include <string>
using namespace std;
class Solution {
public:
    bool isOneEditDistance(string s, string t) {
        int m = s.size();
        int n = t.size();
        if (m > n)
            return isOneEditDistance(t, s);
        if (n - m > 1)
            return false;
        bool mismatch = false;
        for (int i = 0; i < m; i++) {
            if (s[i] != t[i]) {
                mismatch = true;
                if (m == n)
                    s[i] = t[i];
                else
                    s.insert(i, 1, t[i]);
                break;
            }
        }
        return (!mismatch && n - m == 1) || (mismatch && s == t);
    }
};
